support/vm/testing.go

```diff
@@ -0,0 +1,276 @@
+
+	vrState := verifreg.ConstructState(emptyMapCID, rootVerifier.idAddr)
+	initializeActor(ctx, t, vm, vrState, builtin.VerifiedRegistryActorCodeID, builtin.VerifiedRegistryActorAddr, big.Zero())
+
+	// burnt funds
+	initializeActor(ctx, t, vm, &account.State{Address: builtin.BurntFundsActorAddr}, builtin.AccountActorCodeID, builtin.BurntFundsActorAddr, big.Zero())
+
+	// verified root
+	rootVerifierState := &account.State{Address: rootVerifier.pubAddr}
+	initializeActor(ctx, t, vm, rootVerifierState, builtin.AccountActorCodeID, rootVerifier.idAddr, big.Zero())
+
+	_, err = vm.checkpoint()
+	require.NoError(t, err)
+
+	return vm
+}
+
+// Creates n account actors in the VM with the given balance
+func CreateAccounts(ctx context.Context, t *testing.T, vm *VM, n int, balance abi.TokenAmount) []address.Address {
+	var initState initactor.State
+	err := vm.GetState(builtin.InitActorAddr, &initState)
+	require.NoError(t, err)
+
+	addrPairs := make([]addrPair, n)
+	for i := range addrPairs {
+		addr := actor_testing.NewBLSAddr(t, 93837778)
+		idAddr, err := initState.MapAddressToNewID(vm.store, addr)
+		require.NoError(t, err)
+
+		addrPairs[i] = addrPair{
+			pubAddr: addr,
+			idAddr:  idAddr,
+		}
+	}
+	err = vm.setActorState(ctx, builtin.InitActorAddr, &initState)
+	require.NoError(t, err)
+
+	pubAddrs := make([]address.Address, len(addrPairs))
+	for i, addrPair := range addrPairs {
+		st := &account.State{Address: addrPair.pubAddr}
+		initializeActor(ctx, t, vm, st, builtin.AccountActorCodeID, addrPair.idAddr, balance)
+		pubAddrs[i] = addrPair.pubAddr
+	}
+	return pubAddrs
+}
+
+//
+// Invocation expectations
+//
+
+func ExpectObject(v runtime.CBORMarshaler) *objectExpectation {
+	return &objectExpectation{v}
+}
+
+// distinguishes a non-expectation from an expectation of nil
+type objectExpectation struct {
+	val runtime.CBORMarshaler
+}
+
+func ExpectAttoFil(amount big.Int) *big.Int                    { return &amount }
+func ExpectAddress(addr address.Address) *address.Address      { return &addr }
+func ExpectBytes(b []byte) *objectExpectation                  { return ExpectObject(runtime.CBORBytes(b)) }
+func ExpectExitCode(code exitcode.ExitCode) *exitcode.ExitCode { return &code }
+
+// match by cbor encoding to avoid inconsistencies in internal representations of effectively equal objects
+func (oe objectExpectation) matches(obj interface{}) bool {
+	if oe.val == nil || obj == nil {
+		return oe.val == nil && obj == nil
+	}
+
+	paramBuf1 := new(bytes.Buffer)
+	oe.val.MarshalCBOR(paramBuf1) // nolint: errcheck
+	marshaller, ok := obj.(runtime.CBORMarshaler)
+	if !ok {
+		return false
+	}
+	paramBuf2 := new(bytes.Buffer)
+	if marshaller != nil {
+		marshaller.MarshalCBOR(paramBuf2) // nolint: errcheck
+	}
+	return bytes.Equal(paramBuf1.Bytes(), paramBuf2.Bytes())
+}
+
+var okExitCode = exitcode.Ok
+var ExpectOK = &okExitCode
+
+type ExpectInvocation struct {
```

Comment:
Please document this, e.g. the meaning of zero/nil values.

Reply:
✅